src/Features/CSharp/Portable/UseExpressionBodyForLambda/UseExpressionBodyForLambdaCodeStyleProvider.cs

```diff
@@ -0,0 +1,434 @@
+            }
+
+            // They don't have an expression body.  See if we could convert the block they 
+            // have into one.
+            var options = declaration.SyntaxTree.Options;
+            return TryConvertToExpressionBody(declaration, options, preference, out _, out _);
+        }
+
+        public static bool CanOfferUseBlockBody(
+            SemanticModel semanticModel, ExpressionBodyPreference preference,
+            LambdaExpressionSyntax declaration, CancellationToken cancellationToken)
+        {
+            var userPrefersBlockBodies = preference == ExpressionBodyPreference.Never;
+            if (!userPrefersBlockBodies)
+            {
+                // If the user doesn't even want block bodies, then certainly do not offer.
+                return false;
+            }
+
+            var expressionBodyOpt = GetExpressionBody(declaration);
+            if (expressionBodyOpt == null)
+            {
+                // they already have a block body.
+                return false;
+            }
+
+            // We need to know what sort of lambda this is (void returning or not) in order to be
+            // able to create the right sort of block body (i.e. with a return-statement or
+            // expr-statement).  So, if we can't figure out what lambda type this is, we should not
+            // proceed.
+            var lambdaType = semanticModel.GetTypeInfo(declaration, cancellationToken).ConvertedType as INamedTypeSymbol;
+            if (lambdaType == null || lambdaType.DelegateInvokeMethod == null)
+            {
+                return false;
+            }
+
+            var canOffer = expressionBodyOpt.TryConvertToStatement(
+                semicolonTokenOpt: default, createReturnStatementForExpression: false, out _);
+            if (!canOffer)
+            {
+                // Couldn't even convert the expression into statement form.
+                return false;
+            }
+
+            var languageVersion = ((CSharpParseOptions)declaration.SyntaxTree.Options).LanguageVersion;
+            if (expressionBodyOpt.IsKind(SyntaxKind.ThrowExpression) &&
+                languageVersion < LanguageVersion.CSharp7)
+            {
+                // Can't convert this prior to C# 7 because ```a => throw ...``` isn't allowed.
+                return false;
+            }
+
+            return true;
+        }
+
+        private static Location GetDiagnosticLocation(LambdaExpressionSyntax declaration)
+            => Location.Create(declaration.SyntaxTree,
+                    TextSpan.FromBounds(declaration.SpanStart, declaration.ArrowToken.Span.End));
+
+        #endregion
+
+        #region fixing
+
+        protected override Task<ImmutableArray<CodeAction>> ComputeCodeActionsAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            var codeAction = new MyCodeAction(
+                diagnostic.GetMessage() + "-Fix",
```

Comment:
this -Fix suffix is simply for ensuring that the right codepath was generating the code action when testing this in VS.